File: core/src/main/javascript/MslException.js
/**
 * Copyright (c) 2012-2017 Netflix, Inc.  All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Thrown when an exception occurs within the Message Security Layer.
 *
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
(function(require, module) {
	"use strict";
	
	var Class = require('./util/Class.js');
	var MslConstants = require('./MslConstants.js');
	
    var MslException = module.exports = Class.create(new Error());

    var proto = {
        /**
         * Construct a new MSL exception with the specified error, details, and
         * cause.
         *
         * @param {MslError} error the error.
         * @param {string} details the details text. May be null or undefined.
         * @param {Error} cause the cause. May be null or undefined.
         * @constructor
         */
        init: function init(error, details, cause) {
            var self = this;
            
            // Fix my stack trace.
            if (Error.captureStackTrace)
                Error.captureStackTrace(this, this.constructor);

            // Construct the message.
            var message = error.message;
            if (details)
                message += " [" + details + "]";

            // Hide the message ID inside this scope.
            var messageId;
            /**
             * Set the message ID of the message associated with the exception. This
             * does nothing if the message ID is already set.
             *
             * @param id message ID of the message associated with this error.
             */
            function setMessageId(id) {
                if (id < 0 || id > MslConstants.MAX_LONG_VALUE)
                    throw new RangeError("Message ID " + id + " is outside the valid range.");
                if (!getMessageId())
                    messageId = id;
            }

            /**
             * Returns the message ID of the message associated with the exception. May
             * be null if there is no message associated or the exception was thrown
             * before extracting the message ID.
             *
             * @return the message ID or null/undefined.
             */
            function getMessageId() {
                if (messageId)
                    return messageId;
                if (self.cause && self.cause instanceof MslException)
                    return self.cause.messageId;
                return undefined;
            }

            // Construct a better stack trace.
            var originalStack = this.stack;
            function getStack() {
                var trace = self.toString();
                if (originalStack)
                    trace += "\n" + originalStack;
                if (cause && cause.stack)
                    trace += "\nCaused by " + cause.stack;
                return trace;
            }

            // The properties.
            var props = {
                message: { value: message, writable: false, configurable: true },
                error: { value: error, writable: false, configurable: true },
                cause: { value: cause, writable: false, configurable: true },
                name: { value: "MslException", writable: false, configurable: true },
                masterToken: { value: null, writable: true, configurable: false },
                entityAuthenticationData: { value: null, writable: true, configurable: false },
                userIdToken: { value: null, writable: true, configurable: false },
                userAuthenticationData: { value: null, writable: true, configurable: false },
                messageId: { get: getMessageId, set: setMessageId, configurable: true },
                stack: { get: getStack, configurable: true }
            };
            Object.defineProperties(this, props);
        },

        /**
         * Set the entity associated with the exception, using a master token. This
         * does nothing if the entity is already set.
         * 
         * @param {MasterToken} masterToken entity associated with the error. May be null.
         * @return {MslException} this.
         */
        setMasterToken: function setMasterToken(masterToken) {
            if (masterToken && !this.masterToken && !this.entityAuthenticationData)
                this.masterToken = masterToken;
            return this;
        },

        /**
         * Set the entity associated with the exception, using entity
         * authentication data. This does nothing if the entity is already set.
         *
         * @param {EntityAuthenticationData} entityAuthData entity associated with the error. May be null.
         * @return {MslException} this.
         */
        setEntityAuthenticationData: function setEntityAuthenticationData(entityAuthData) {
            if (entityAuthData && !this.masterToken && !this.entityAuthenticationData)
                this.entityAuthenticationData = entityAuthData;
            return this;
        },

        /**
         * Set the user associated with the exception, using a user ID token. This
         * does nothing if the user is already set.
         *
         * @param {UserIdToken} userIdToken the user ID token associated with the error. May be null.
         * @return {MslException} this.
         */
        setUserIdToken: function setUserIdToken(userIdToken) {
            if (userIdToken && !this.userIdToken && !this.userAuthenticationData)
                this.userIdToken = userIdToken;
            return this;
        },

        /**
         * Set the user associated with the exception, using user authentication
         * data. This does nothing if the user is already set.
         *
         * @param {UserAuthenticationData} userAuthData the user authentication data associated with the error. May be null.
         * @return {MslException} this.
         */
        setUserAuthenticationData: function setUserAuthenticationData(userAuthData) {
            if (userAuthData && !this.userIdToken && !this.userAuthenticationData)
                this.userAuthenticationData = userAuthData;
            return this;
        },

        /**
         * Set the message ID of the message associated with the exception. This
         * does nothing if the message ID is already set.
         *
         * @param {number} id message ID of the message associated with this error.
         * @return {MslException} this.
         */
        setMessageId: function setMessageId(id) {
            this.messageId = id;
            return this;
        },

        /**
         * @return a string containing the exception type and message.
         */
        toString: function toString() {
            return this.name + ': ' + this.message;
        }
    };

    // Attach methods.
    MslException.mixin(proto);
})(require, (typeof module !== 'undefined') ? module : mkmodule('MslException'));

File: core/src/main/cpp/MslException.cpp
/**
 * Copyright (c) 2016-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <MslException.h>
#include <tokens/UserIdToken.h>
#include <userauth/UserAuthenticationData.h>
#include <sstream>

using namespace std;

namespace netflix {
namespace msl {

namespace {

// Recurse through the cause exception stack, until we find the first
// MslException (typename D) that has a non-empty func() return, or else
// return the default of whatever func() wants to return.
template<typename T, typename D>
T recurseCauses(shared_ptr<IException> cause, T (D::*func)() const) {
    if (cause && instanceof<D>(cause.get())) {
        const shared_ptr<D> mslCause = dynamic_pointer_cast<D>(cause);
        return (mslCause.get()->*func)();  // recurse
    } else {
        return T();
    }
}

} // namespace anonymous

MslException::MslException(const MslError& error)
    : Exception(error.getMessage()), error_(error), messageId_(0)
{
}

MslException::MslException(const MslError& error, const string& details)
    : Exception(error.getMessage() + " [" + details + "]"),
      error_(error),
      messageId_(0)
{
}

MslException::MslException(const MslError& error, const string& details,
        const IException& cause)
    : Exception(error.getMessage() + " [" + details + "]", cause),
      error_(error),
      messageId_(0)
{
}

MslException::MslException(const MslError& error, const IException& cause)
    : Exception(error.getMessage(), cause), error_(error), messageId_(0)
{
}

int64_t MslException::getMessageId() const
{
    if (messageId_ != 0)
        return messageId_;
    return recurseCauses(cause_, &MslException::getMessageId);
}

MslException& MslException::setMessageId(int64_t messageId)
{
    if (messageId > MslConstants::MAX_LONG_VALUE)
    {
        ostringstream sstream;
        sstream << "Message ID " << messageId << " is outside the valid range.";
        throw out_of_range(sstream.str());
    }
    if (getMessageId() == 0ll)
        messageId_ = messageId;
    return *this;
}

shared_ptr<tokens::MasterToken> MslException::getMasterToken() const
{
    if (masterToken_)
        return masterToken_;
    return recurseCauses(cause_, &MslException::getMasterToken);
}

MslException& MslException::setMasterToken(shared_ptr<tokens::MasterToken> masterToken)
{
    if (!getMasterToken() && !getEntityAuthenticationData())
        masterToken_ = masterToken;
    return *this;
}

std::shared_ptr<entityauth::EntityAuthenticationData> MslException::getEntityAuthenticationData() const
{
    if (entityAuthData_)
        return entityAuthData_;
    return recurseCauses(cause_, &MslException::getEntityAuthenticationData);
}

MslException& MslException::setEntityAuthenticationData(shared_ptr<entityauth::EntityAuthenticationData> entityAuthData)
{
    if (!getMasterToken() && !getEntityAuthenticationData())
        entityAuthData_ = entityAuthData;
    return *this;
}

MslException& MslException::setUserIdToken(shared_ptr<tokens::UserIdToken> userIdToken)
{
    if (!getUserIdToken() && !getUserAuthenticationData())
        userIdToken_ = userIdToken;
    return *this;
}

shared_ptr<tokens::UserIdToken> MslException::getUserIdToken() const
{
    if (userIdToken_)
        return userIdToken_;
    return recurseCauses(cause_, &MslException::getUserIdToken);
}

MslException& MslException::setUserAuthenticationData(shared_ptr<userauth::UserAuthenticationData> userAuthData)
{
    if (!getUserIdToken() && !getUserAuthenticationData())
        userAuthData_ = userAuthData;
    return *this;
}

std::shared_ptr<userauth::UserAuthenticationData> MslException::getUserAuthenticationData() const
{
    if (userAuthData_)
        return userAuthData_;
    return recurseCauses(cause_, &MslException::getUserAuthenticationData);
}

} /* namespace msl */
} /* namespace netflix */
